mindsdb/interfaces/model/model_controller.py

```diff
@@ -1,366 +1,383 @@
+        df = ds.df
+
+        print(problem_definition, df)
+        json_ai = lightwood.json_ai_from_problem(df, problem_definition)
+        code = lightwood.code_from_json_ai(json_ai)
+
+        db_p = db.Predictor(
+            company_id=company_id,
+            name=name,
+            json_ai=json_ai.to_dict(),
+            code=code,
+            datasource_id=datasource_id,
+            mindsdb_version=mindsdb_version,
+            lightwood_version=lightwood_version,
+            to_predict=[problem_definition.target],
+            data={'status': 'untrained', 'name': name}
+        )
+        db.session.add(db_p)
+        db.session.commit()
+        delete_process_mark('learn')
+        print('generate predicrtor end')
+
+    def edit_json_ai(self, name: str, json_ai: dict, company_id=None):
+        """Edit an existing predictor's json_ai"""
+
+        predictor_record = db.session.query(db.Predictor).filter_by(company_id=company_id, name=name).first()
+        assert predictor_record is not None
+
+        try:
+            json_ai: lightwood.JsonAI = lightwood.JsonAI.from_dict(json_ai)  # type: ignore
+            code = lightwood.code_from_json_ai(json_ai)
+        except Exception as e:
+            print(f'Failed to generate predictor from json_ai: {e}')
+            return False
+        else:
+            predictor_record.code = code
+            predictor_record.code = json_ai
+            db.session.commit()
+            return True
+
+    def edit_code(self, name: str, code: str, company_id=None):
+        """Edit an existing predictor's code"""
+
+        predictor_record = db.session.query(db.Predictor).filter_by(company_id=company_id, name=name).first()
+        assert predictor_record is not None
+        
+        try:
+            # TODO: make this safe from code injection (on lightwood side)
+            lightwood.predictor_from_code(code)
+        except Exception as e:
+            print(f'Failed to generate predictor from json_ai: {e}')
+            return False
+        else:
+            predictor_record.code = code
+            predictor_record.json_ai = None
+            db.session.commit()
+            return True
+
+    def fit_predictor(self, name: str, from_data: dict, join_learn_process: bool, company_id: int):
+        create_process_mark('learn')
+        print('fit predicrtor start')
+        """Train an existing predictor"""
+
+        ds_cls = getattr(mindsdb_datasources, from_data['class'])
+        ds = ds_cls(*from_data['args'], **from_data['kwargs'])
+        df = ds.df
+
+        predictor_record = db.session.query(db.Predictor).filter_by(company_id=company_id, name=name).first()
+        assert predictor_record is not None
+        predictor_record.data = {'status': 'training', 'name': name}
+        db.session.commit()
+
+        p = LearnProcess(predictor_record.id, df)
+        p.start()
+        if join_learn_process:
+            p.join()
```

Comment:
need add `delete_process_mark('learn')` to the end

Reply:
Should already be happening in the learn process so not needed